src/charm.py

```diff
@@ -75,63 +73,55 @@ def _handle_event(self, event: ops.EventBase) -> None:  # noqa: C901
             self._ensure_installed()
             config = self._get_config()
             self._mount_filesystems(config)
-        except StopCharmError:
+        except StopCharmError as e:
             # This was the cleanest way to ensure the inner methods can still return prematurely
             # when an error occurs.
+            self.app.status = e.status
             return
 
-        self.unit.status = ops.ActiveStatus("Mounted shares.")
+        self.unit.status = ops.ActiveStatus("Mounted filesystems.")
 
     def _ensure_installed(self):
         """Ensure the required packages are installed into the unit."""
         if not self._mounts_manager.installed:
             self.unit.status = ops.MaintenanceStatus("Installing required packages.")
-            self._mounts_manager.ensure(apt.PackageState.Present)
+            self._mounts_manager.install()
 
     def _get_config(self) -> dict[str, dict[str, str | bool]]:
         """Get and validate the configuration of the charm."""
         try:
-            config = json.loads(str(self.config.get("mountinfo", "")))
+            config = json.loads(str(self.config.get("mounts", "")))
             validate(config, CONFIG_SCHEMA)
             config: dict[str, dict[str, str | bool]] = config
             for fs, opts in config.items():
                 for opt in ["noexec", "nosuid", "nodev", "read-only"]:
                     opts[opt] = opts.get(opt, False)
             return config
         except (json.JSONDecodeError, ValidationError) as e:
-            self.app.status = ops.BlockedStatus(
-                f"invalid configuration for option `mountinfo`. reason: {e}"
+            raise StopCharmError(
+                ops.BlockedStatus(f"invalid configuration for option `mounts`. reason:\n{e}")
             )
-            raise StopCharmError()
 
     def _mount_filesystems(self, config: dict[str, dict[str, str | bool]]):
         """Mount all available filesystems for the charm."""
-        shares = self._fs_share.endpoints
-        active_filesystems = set()
-        for fs_type, count in Counter([share.fs_info.fs_type() for share in shares]).items():
+        endpoints = self._filesystems.endpoints
+        for fs_type, count in Counter(
+            [endpoint.info.filesystem_type() for endpoint in endpoints]
+        ).items():
             if count > 1:
-                self.app.status = ops.BlockedStatus(
-                    f"Too many relations for mount type `{fs_type}`."
+                raise StopCharmError(
+                    ops.BlockedStatus(f"Too many relations for mount type `{fs_type}`.")
```

Comment:
This goes away if we use a single application per filesystem, which I think is a better option for the time being. There's a lot that needs to be unpacked/designed/figured out to support multiple filesystems cleanly (e.g. how to do the local mountpoint to remote fs mapping without having the user know lots of details about the filesystem servers ahead of time).